### PR TITLE
Feature/sprint ioda converters ne viirs modis l2 oc2ioda

### LINKED ISSUE #870: Upgrade converter for JPSS VIIRS/MODIS ocean color Level2 to follow the JEDI Data Conventions
Update the converter identified in the title (hereafter referred to as the **viirs_modis_l2_oc2ioda.py**) to follow the JEDI Data Conventions.

- Update variable names
- Ensure variable units match the units in the conventions table
- Update the global attributes dictionary as needed
- Run the converter test to make sure it runs (but will fail because of unupdated test reference)
- Run ```ioda-validate.x``` on your output file and verify that there are no errors or warnings
- Update test reference file
- Run unit tests making use of the target converter and verify that they pass
- Make sure you verify the results by examining the file using ```ncdump -t```, ```h5dump```, ```HDFview```, or a similar application.

## ZenHub organization

Please make sure that you have properly set:
- Labels: OBS3
- Estimate: An estimate of how long this issue will take to resolve. 1 = half a day. 2 = full day. 3 = two days, 5 = three days.
- Epics: [Sprint: Implement the JEDI Data Conventions in the Converters](https://github.com/JCSDA-internal/ioda-converters/issues/813)
- Assignees: Anyone **working** on this issue. Anyone who is not working directly on the issue, but
  who should receive notifications, should be listed below.
- Reviewers

## Notifications

Fill in a list of GitHub handles here. People here are not *assigned*, but instead receive notifications
about this issue's progress.

Notifications should always go to: @rhoneyager.

Other people:

### PR DESCRIPTION
**Pull request by nicholasesposito**
## Description

This updates the python HDF5 converter to the latest IODA conventions

### Issue(s) addressed

- fixes #870 

## Acceptance Criteria (Definition of Done)

successful pass of the ctest 'test_iodaconv_viirs_jpss1_oc_l2' , 'test_iodaconv_modis_aqua_oc_l2', 

## Dependencies

##NOTE
oceanMassContentOfParticulateOrganicMatterExpressedAsCarbon and massConcentrationOfChlorophyllInSeaWater are not yet in ObsSpace.yaml, so they still come up as warnings when using ioda-validate.x

### CODE REVIEW

**Comment by rhoneyager** [comment]
oceanMassContentOfParticulateOrganicMatterExpressedAsCarbon and massConcentrationOfChlorophyllInSeaWater will be discussed in the conventions meeting today.

File: src/marine/argoClim2ioda.py [hunk 1]
```diff
@@ -222,13 +229,13 @@ def main():
 
     argo = argoClim(filename, begindate=bdate, enddate=edate)
 
-    varDict = {
-        'ob_tem': 'temperature',
-    }
+#    varDict = {
```
Comment (srherbener) [inline src/marine/argoClim2ioda.py:232]: Please remove commented code. Thanks!

File: src/marine/viirs_modis_l2_oc2ioda.py [hunk 2]
```diff
@@ -238,47 +241,47 @@ def main():
...
     # prepare global attributes we want to output in the file,
     # in addition to the ones already loaded in from the input file
-    GlobalAttrs['date_time_string'] = args.date.strftime("%Y-%m-%dT%H:%M:%SZ")
+    GlobalAttrs['datetimeReference'] = args.date.strftime("%Y-%m-%dT%H:%M:%SZ")
     GlobalAttrs['thinning'] = args.thin
     GlobalAttrs['converter'] = os.path.basename(__file__)
-    DimDict['nlocs'] = nlocs
-    GlobalAttrs['nlocs'] = np.int32(DimDict['nlocs'])
+    DimDict['Location'] = Location
+#    GlobalAttrs['Location'] = np.int32(DimDict['Location'])
```
Comment (srherbener) [inline src/marine/viirs_modis_l2_oc2ioda.py:253]: Please remove commented code. Thanks!

File: src/marine/viirs_modis_l2_oc2ioda.py [hunk 3]
```diff
@@ -238,47 +241,47 @@ def main():
...
     if args.chl:
         VarAttrs[output_var_names[1], global_config['oval_name']]['units'] = \
-            'mg ^m-3'
+            'mg m-3'
         VarAttrs[output_var_names[1], global_config['oerr_name']]['units'] = \
-            'mg ^m-3'
-        VarAttrs[output_var_names[1], global_config['opqc_name']]['units'] = \
-            'unitless'
+            'mg m-3'
+#        VarAttrs[output_var_names[1], global_config['opqc_name']]['units'] = \
```
Comment (srherbener) [inline src/marine/viirs_modis_l2_oc2ioda.py:276]: Please remove commented code. Thanks!

**Comment by nicholasesposito** [comment]
I believe I made the correct edits. I may have been on the incorrect branch
so let me know if things look funky. I think it's good to go.

Sincerely,
Nick

-----
Nicholas Esposito (he/him/his)
I.M. Systems Group, Scientific Programmer / Analyst
***@***.***/NCEP/EMC
5830 University Research Court, Room 2058
College Park, MD 20740

Email: ***@***.***
Office Phone: (301)683-3706

**Review by srherbener — APPROVED** [review]
Looks good. Thanks!

**Review by CoryMartin-NOAA — APPROVED** [review]

**Comment by CoryMartin-NOAA** [comment]
Code looks good, need to merge in staging branch and resolve conflicts.

**Comment by srherbener** [comment]
> @rhoneyager @srherbener can you merge develop into the feature/sprint-ioda-converters branch? It seems there are changes reflected here that are due to recent changes in develop.

Done! Sorry about the feature/sprint-ioda-converters branch getting out of sync.

**Comment by srherbener** [comment]
Please resolve the conflict, and then we can check if the merge from develop into feature/sprint-ioda-converters fixed the issue. Thanks!

**Comment by srherbener** [comment]
Thanks for fixing the conflicts

**Comment by srherbener** [comment]
Looks like the unit tests for MODIS AQUA OC and VIIRS JPSS1 OC are passing, but the validator is failing on the new output files for these converters. I can merge once these validator errors are fixed.

Does fixing these entail merging in more PRs that update the ObsSpace.yaml file?

**Comment by nicholasesposito** [comment]
The validator is failing as a result of warnings, but not errors. For example:
  Variable ObsValue/oceanMassContentOfParticulateOrganicMatterExpressedAsCarbon
   Warning: Variable 'ObsValue/oceanMassContentOfParticulateOrganicMatterExpressedAsCarbon' is not listed in the YAML conventions file.
  Variable ObsValue/massConcentrationOfChlorophyllInSeaWater
   Warning: Variable 'ObsValue/massConcentrationOfChlorophyllInSeaWater' is not listed in the YAML conventions file.

I imagine adding them would allow the validator to pass.

**Comment by srherbener** [comment]
Still seeing test failures for modis aqua oc, and viirs jpss1 oc tests and validator warnings. I'm okay with leaving the review approvals intact, but these test failures need to be fixed before merging. Thanks!

**Comment by nicholasesposito** [comment]
OK new updates:  I just opened an issue and PR to update ObsSpace.yaml with variables needed to get the tests to pass.
https://github.com/JCSDA-internal/ioda/issues/693
https://github.com/JCSDA-internal/ioda/pull/694

Once those are merged. The tests should pass:
  Test #452: test_iodaconv_viirs_jpss1_oc_l2
  Test #453: test_iodaconv_modis_aqua_oc_l2
  Test #586: iodaconv_validate_testoutput_modis_aqua_oc_l2.nc
  Test #617: iodaconv_validate_testoutput_viirs_jpss1_oc_l2.nc

Once the tests pass. This may be merged and closed.

**Comment by srherbener** [comment]
Out of the list of tests expected to pass, the validator tests pass but the converter tests do not. CDash is reporting similar errors for the converter tests. Here's the modis aqua report:
```
Reading  testinput/modis_aqua_oc_l2.nc
DIFFER : NAMES OF RECORDS : Location <> 
DIFFER : LENGTHS OF RECORDS : Location (3637) <>  (0)
DIFFER : VARIABLE : Location : TYPE : INT64 <> INT
DIFFER : VARIABLE : Location : DIMENSION Location IS RECORD IN FILE "testrun/modis_aqua_oc_l2.nc" BUT NOT IN "testoutput/modis_aqua_oc_l2.nc"
/home/jedi/bin/iodaconv_comp.sh: line 20:  7236 Segmentation fault      (core dumped) nccmp testrun/$file_name testoutput/$file_name -d -m -g -f -S -T ${tol}
```

I'm trying this branch on my Mac to see what happens.

**Comment by srherbener** [comment]
I see the same failures on my Mac. It looks like the test reference files have the dimension Location with a fixed size and single precision integer, while the test run files have Location with unlimited size and and double precision integer. I think these differences are confusing the nccmp tool which ultimately segfaults after listing a few differences it detected.

If you are certain that the scripts are producing the correct output, I think the fix is to simply update the test reference files.

**Comment by nicholasesposito** [comment]
It's weird because the results in testoutput are what i get when I run the code on Hera. Perhaps it's an environment issue? Is there someone that has experience with weird results like this? Any ideas on what could be done?

**Comment by srherbener** [comment]
> It's weird because the results in testoutput are what i get when I run the code on Hera. Perhaps it's an environment issue? Is there someone that has experience with weird results like this? Any ideas on what could be done?

The results on Hera are indeed strange. I'm not sure why they would come out different than other platforms. What happens if you run on Orion? If you don't have access to Orion, I could give it a try.

I don't have access to Hera so if we need to debug there we will need someone to help. Perhaps @CoryMartin-NOAA could help with Hera.

**Comment by srherbener** [comment]
I tried Orion and the tests pass there so the results are the same as Hera. I did an ncdump on the `testrun/modis_aqua_oc_l2.nc4` files (ie what was produced by the converter during the test) from both Orion and my Mac. (The textual output of ncdump is called CDL.) Then I diff'd the two CDL files:
```
bash-5.1$ diff modis_aqua.mac.cdl modis_aqua.orion.cdl
3c3
< 	Location = UNLIMITED ; // (3637 currently)
---
> 	Location = 3637 ;
5c5
< 	int64 Location(Location) ;
---
> 	int Location(Location) ;
174c174
<   		dateTime:_FillValue = 253370764800000000LL ;
---
>   		dateTime:_FillValue = -4883652231933722624LL ;
```

So the Mac creates Location as an int64 with unlimited size, whereas Orion creates Location as an int with fixed size. Everything else in the files match. The idoa_conv_engines.py library is trying to setup Location as an int64 with unlimited size so it looks like the Mac is getting the intended result.

Any ideas why Orion and Hera are creating Location as int (single precision) with a fixed size?

**Comment by srherbener** [comment]
> I tried Orion and the tests pass there so the results are the same as Hera. I did an ncdump on the `testrun/modis_aqua_oc_l2.nc4` files (ie what was produced by the converter during the test) from both Orion and my Mac. (The textual output of ncdump is called CDL.) Then I diff'd the two CDL files:
> 
> ```
> bash-5.1$ diff modis_aqua.mac.cdl modis_aqua.orion.cdl
> 3c3
> < 	Location = UNLIMITED ; // (3637 currently)
> ---
> > 	Location = 3637 ;
> 5c5
> < 	int64 Location(Location) ;
> ---
> > 	int Location(Location) ;
> 174c174
> <   		dateTime:_FillValue = 253370764800000000LL ;
> ---
> >   		dateTime:_FillValue = -4883652231933722624LL ;
> ```
> 
> So the Mac creates Location as an int64 with unlimited size, whereas Orion creates Location as an int with fixed size. Everything else in the files match. The idoa_conv_engines.py library is trying to setup Location as an int64 with unlimited size so it looks like the Mac is getting the intended result.
> 
> Any ideas why Orion and Hera are creating Location as int (single precision) with a fixed size?

@nicholasesposito, I just rebuilt Orion from scratch (ie, wiped out my build area), and now I'm seeing matching results with my Mac. The Location dimension is int64 and unlimited size.

I suspect the issue was that the ioda python modules got out of sync and didn't get updated with the latest versions. (These modules are where the int64 and unlimited size is being set for Location.) And I'm wondering if the same thing has happened with your build area on Hera.

Please try wiping out your build are on Hera and rebuilding to see if that gets things consistent (Location is int64, unlimited size) with the other platforms.

Thanks!

**Comment by nicholasesposito** [comment]
Steve. Thanks for trying on Orion. You mentioned that maybe it's the python modules that are the issue. What is the environment/python version you are using? These are the modules I have loaded. How similar is this to what you have?
```

  1) git-lfs/2.11.0      9) jedi-impi/2020.2      17) pio/2.5.1-debug      25) json/3.9.1
  2) intel/2020.2       10) hdf5/1.12.0           18) cmake/3.20.1         26) json-schema-validator/2.1.0
  3) jedi-intel/2020.2  11) pnetcdf/1.12.2        19) ecbuild/ecmwf-3.6.1  27) hpc/1.2.0
  4) szip/2.1.1         12) netcdf/4.7.4          20) eckit/ecmwf-1.16.0   28) miniconda3/4.6.14
  5) zlib/1.2.11        13) boost-headers/1.68.0  21) fckit/ecmwf-0.9.2    29) iodaconv/1.0.0
  6) udunits/2.2.28     14) eigen/3.3.7           22) atlas/ecmwf-0.24.1   30) jedi-iodaconv/0.1
  7) gsl_lite/0.37.0    15) bufr/noaa-emc-11.5.0  23) nco/4.9.1
  8) impi/2020.2        16) nccmp/1.8.7.0         24) pybind11/2.7.0
```

**Comment by nicholasesposito** [comment]
I built from scratch and the results were different from earlier. You were definitely right. Location is now int64 and unlimited. Check and let me know! @srherbener

**Comment by srherbener** [comment]
All of the relevant tests are passing now. Thanks for fixing this! 

I have discovered a bug with the fill values on dateTime (due to a difference I see on my Mac), but all of the converters in the feature/sprint-ioda-converters are being impacted by this bug. I'll merge this now and submit a ZenHub issue for the bug I found.

**Pull request merged**